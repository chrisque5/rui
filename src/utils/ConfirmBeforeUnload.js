import { useEffect, useCallback } from 'react';

const ConfirmBeforeUnload = () => {
  const onBeforeUnload = useCallback((e) => {
    e.preventDefault();
    e.returnValue = '';
  }, []);

  useEffect(() => {
    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [onBeforeUnload]);

  return null;
};

export default ConfirmBeforeUnload;
