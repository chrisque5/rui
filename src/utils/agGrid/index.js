import {
  approvalsCellRenderer, approvalsSummaryCellRenderer, isUnderInvestigationCellRenderer, stpCellRenderer, stpFilterCellRenderer,
} from './js-renderers';
import {
  dateFormatter, formatDate, numberFormatter, timeFormatter,
  valueDateFormatter, fixingDateFormatter, publishDateFormatter,
  priceFormatter, brokerageFormatter,
} from './formatters';
import { dateFilterComparator, numberFilterComparitor, timeFilterComparator } from './comparitors';
import { getExportMenuItems } from './exportUtils';

export {
  approvalsCellRenderer,
  approvalsSummaryCellRenderer,
  dateFilterComparator,
  dateFormatter,
  formatDate,
  getExportMenuItems,
  isUnderInvestigationCellRenderer,
  numberFilterComparitor,
  numberFormatter,
  stpCellRenderer,
  stpFilterCellRenderer,
  timeFilterComparator,
  timeFormatter,
  valueDateFormatter,
  fixingDateFormatter,
  publishDateFormatter,
  priceFormatter,
  brokerageFormatter,
};
