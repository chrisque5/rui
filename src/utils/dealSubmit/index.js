import mapFormToFwdChain from './mapFormToFwdChain';
import mapFormToNdfChain from './mapFormToNdfChain';
import mapFormToSptChain from './mapFormToSptChain';

export { mapFormToFwdChain, mapFormToNdfChain, mapFormToSptChain };
