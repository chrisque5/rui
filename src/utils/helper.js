import moment from 'moment';
import * as math from 'mathjs';
import { Modal } from 'antd';

export const isValidTerm = (termToTest, selections) => {
  if (termToTest.length < 2) return false;

  const lowerCaseTermToTest = termToTest.toLowerCase();

  const endsWithPeriod = (term) => {
    const periods = ['d', 'w', 'm', 'y'];
    return (periods.includes(term.charAt(term.length - 1)));
  };

  const firstCharacterIsNumber = (term) => {
    const result = term.substring(0, term.length - 1).match(/^[0-9]*$/);
    return result ? result.length > 0 : false;
  };

  const isStaticSelectionValid = (term) => {
    const validStaticTerms = selections.map((x) => x.toLowerCase());
    if (validStaticTerms.includes(term)) {
      return true;
    }

    if (term.startsWith('bmf') || term.startsWith('imm')) {
      let termDaysOnly = term.replace('bmf', '');
      termDaysOnly = termDaysOnly.replace('imm', '');
      const beforeConversion = termDaysOnly;
      termDaysOnly = parseInt(termDaysOnly, 10);
      if (beforeConversion.length === termDaysOnly.toString().length) {
        return (termDaysOnly % 1 === 0 && termDaysOnly > 0 && termDaysOnly < 101);
      }
    }
    return false;
  };

  return isStaticSelectionValid(lowerCaseTermToTest)
    || (endsWithPeriod(lowerCaseTermToTest) && firstCharacterIsNumber(lowerCaseTermToTest));
};

export const doesArrayExistWithValue = (array, index = 0) => {
  if (Array.isArray(array)) {
    if (array && array.length > 0 && array[index]) {
      return true;
    }
  }
  return false;
};

export const setTodayOnNull = (form, id, value) => {
  if (!value) {
    setTimeout(() => {
      form.setFieldsValue({ [id]: moment() });
    });
  }
};

export const dayDifferenceIgnoreHours = (earlierDateMoment, laterDateMoment) => {
  const date1 = earlierDateMoment.clone().startOf('day');
  const date2 = laterDateMoment.clone().startOf('day');
  const dayCount = date1.diff(date2, 'days');

  return dayCount;
};

export const getDateInLocalTZ = () => moment().tz(moment.tz.guess(true)).format('YYYY-MM-DD');
export const getDateTimeInLocalTZ = () => moment().tz(moment.tz.guess(true));

export const parseValueAsBigNumber = (value) => {
  try {
    return math.bignumber(value);
  } catch (err) {
    return null;
  }
};

export const formatAmount = (val) => {
  const inputValue = val.replace(/\s/g, '');
  let multiplier = 1;
  const oneM = 1000000;
  const oneB = 1000000000;
  let amt = inputValue;

  if (Number.isNaN(Number(inputValue)) && inputValue.length > 1) {
    const lastChar = inputValue.charAt(inputValue.length - 1).toLowerCase();
    amt = Number(inputValue.substring(0, inputValue.length - 1));
    if (lastChar === 'm') {
      multiplier = oneM;
    } else if (lastChar === 'b') {
      multiplier = oneB;
    } else {
      return inputValue;
    }
  } else if (Number(inputValue) < oneM) {
    multiplier = oneM;
  }

  const calculatedAmount = math.multiply(math.bignumber(amt), math.bignumber(multiplier));
  return calculatedAmount.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

export const isValidDate = (date, dateFormat) => moment(date, dateFormat || 'L', true).isValid();

export const findCurrencyPair = (ccy1, ccy2, currencies) => (
  currencies.find(
    (currency) => currency.baseCurrency === ccy1 && currency.counterCurrency === ccy2,
  )
);

export const filterSpotExecutionVenues = (execVenues) => execVenues.filter((i) => i.venueId === 'XOFF');

export const isDealTypeSame = (dealType1, dealType2) => (dealType1 === dealType2);

export const isStrategyTypeSame = (strategyType1, strategyType2) => (strategyType1 === strategyType2);

export const ConfirmUnsavedChanges = (message, onYes, onNo) => {
  Modal.confirm({
    title: 'Confirm',
    content: message,
    okText: 'Yes',
    cancelText: 'No',
    onOk: () => {
      if (onYes) {
        setTimeout(() => {
          onYes();
        });
      }
    },
    onCancel: () => {
      if (onNo) {
        setTimeout(() => {
          onNo();
        });
      }
    },
  });
};

export const isObjectEmpty = (obj) => Object.keys(obj).length === 0 && obj.constructor === Object;

export const getUniqueID = () => {
  const s4 = () => Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
  return `${s4() + s4()}-${s4()}`;
};

export const showPreviousValueToolTip = (currentValue, previousValue) => {
  if (currentValue !== previousValue) {
    return { title: `Previous Value: ${previousValue || 'N/A'}`, className: 'deal-input-dirty' };
  }
  return { title: '', className: '' };
};
