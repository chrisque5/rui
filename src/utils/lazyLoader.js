import { lazy } from 'react';

export const lazyWithPreload = (factory) => {
  const Component = lazy(factory);
  Component.preload = factory;
  return Component;
};

export const dummy = 'dummy';
