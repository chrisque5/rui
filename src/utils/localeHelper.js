import enGB from 'antd/lib/locale/en_GB';
import enUS from 'antd/lib/locale/en_US';

export const getLocale = () => {
  const lang = navigator.language || navigator.languages[0] || 'en-GB';
  if (lang === 'en-US') {
    return enUS;
  }
  return enGB;
};

export const dummy = 'dummy';
