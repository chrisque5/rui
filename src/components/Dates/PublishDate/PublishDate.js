import { useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { Form, DatePicker } from 'antd';
import { LoadingOutlined } from '@ant-design/icons';
import moment from 'moment';
import { setTodayOnNull } from '../../../utils/helper';
import { getIsDateResponsePending } from '../../../utils/selectors';

const spinIcon = <LoadingOutlined spin />;

const PublishDate = ({
  form, id, label, testId,
}) => {
  const isPending = useSelector(getIsDateResponsePending);

  return (
    <Form.Item
      label={label}
      className="item-publish-date"
      name={id}
      rules={[{ required: true, message: 'Publish date required' }]}
      initialValue={moment()}
    >
      <DatePicker
        suffixIcon={isPending && spinIcon}
        disabled={isPending}
        data-testid={testId}
        format="L"
        allowClear={false}
        onChange={(value) => setTodayOnNull(form, id, value)}
        tabIndex="-1"
        style={{ width: '100%' }}
      />
    </Form.Item>
  );
};

PublishDate.defaultProps = {
  label: '',
};

PublishDate.propTypes = {
  form: PropTypes.shape().isRequired,
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  testId: PropTypes.string.isRequired,
};

export default PublishDate;
