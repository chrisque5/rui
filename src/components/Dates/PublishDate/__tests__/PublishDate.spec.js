import moment from 'moment';
import {
  renderWithForm, waitFor, getByTitle, fireEvent,
} from 'test-utils/testUtils';
import * as helper from '../../../../utils/helper';
import PublishDate from '../PublishDate';

jest.mock('../../../../utils/helper', () => ({ setTodayOnNull: jest.fn() }));
jest.mock('moment', () => () => jest.requireActual('moment')('2020-04-04T00:00:00.000Z'));

const setTodayOnNullSpy = jest.spyOn(helper, 'setTodayOnNull');

describe('<PublishDate/>', () => {
  const mockForm = {
    resetFields: () => { },
  };

  const defaultProps = {
    form: mockForm,
    id: 'fixingDate1',
    initialValue: moment(),
    testId: 'dtpPublishDate1',
    label: 'Publish Date',
  };

  test('verify snapshot', () => {
    const { container } = renderWithForm(<PublishDate {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  test('verify change event', async () => {
    const { getByTestId } = renderWithForm(<PublishDate {...defaultProps} />);

    await (async () => {
      const dateInput = await waitFor(() => getByTestId('dtpPublishDate1'));
      fireEvent.mouseDown(dateInput);

      const selectedDateElem = await waitFor(() => getByTitle(document.body, '2020-04-11'));
      fireEvent.click(selectedDateElem);
      expect(setTodayOnNullSpy).toHaveBeenCalled();
    });
  });
});
