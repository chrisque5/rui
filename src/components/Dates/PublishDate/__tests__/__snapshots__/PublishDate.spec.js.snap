// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PublishDate/> verify snapshot 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-row ant-form-item item-publish-date"
    >
      <div
        class="ant-col ant-form-item-label"
      >
        <label
          class="ant-form-item-required"
          for="fixingDate1"
          title="Publish Date"
        >
          Publish Date
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-picker"
              style="width: 100%;"
            >
              <div
                class="ant-picker-input"
              >
                <input
                  autocomplete="off"
                  data-testid="dtpPublishDate1"
                  id="fixingDate1"
                  placeholder="Select date"
                  readonly=""
                  size="12"
                  tabindex="-1"
                  title="04/04/2020"
                  value="04/04/2020"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
